File: src/lib/terminal.ts
/**
 * Thin wrappers around Tauri `invoke` / `listen` for PTY session management.
 *
 * Each function maps 1:1 to a Rust `#[tauri::command]` handler. Errors are
 * propagated as rejected promises; callers are responsible for catch/logging.
 */

import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type { BackendCapabilities, BackendType } from "./terminalTheme";

/**
 * Spawns a new PTY shell session on the backend.
 * @param cwd - Starting working directory; when omitted the backend uses its default.
 * @param env - Environment variables to pass to the shell process. These are inherited
 *   by all child processes (including Claude CLI → MCP server). MAESTRO_SESSION_ID is
 *   automatically set by the backend.
 * @returns The numeric session ID assigned by the backend.
 */
export async function spawnShell(cwd?: string, env?: Record<string, string>): Promise<number> {
  return invoke<number>("spawn_shell", { cwd: cwd ?? null, env: env ?? null });
}

/** Writes raw bytes to the PTY stdin of the given session. */
export async function writeStdin(sessionId: number, data: string): Promise<void> {
  return invoke("write_stdin", { sessionId, data });
}

/** Notifies the backend PTY of a terminal dimension change (rows x cols). */
export async function resizePty(sessionId: number, rows: number, cols: number): Promise<void> {
  return invoke("resize_pty", { sessionId, rows, cols });
}

/** Terminates the backend PTY process and cleans up the session. */
export async function killSession(sessionId: number): Promise<void> {
  return invoke("kill_session", { sessionId });
}

/** AI mode variants matching the backend enum. */
export type AiMode = "Claude" | "Gemini" | "Codex" | "Plain";

/** CLI command configuration for each AI mode */
export const AI_CLI_CONFIG: Record<AiMode, { command: string | null; installHint: string }> = {
  Claude: {
    command: "claude",
    installHint: "npm install -g @anthropic-ai/claude-code"
  },
  Gemini: {
    command: "gemini",
    installHint: "npm install -g @google/gemini-cli"
  },
  Codex: {
    command: "codex",
    installHint: "npm install -g codex"
  },
  Plain: {
    command: null,
    installHint: ""
  },
};

/** Checks if a CLI tool is available in the user's PATH */
export async function checkCliAvailable(command: string): Promise<boolean> {
  return invoke<boolean>("check_cli_available", { command });
}

/** Session config returned by createSession. */
export interface SessionConfig {
  id: number;
  mode: AiMode;
  branch: string | null;
  status: string;
  worktree_path: string | null;
  project_path: string;
}

/** Creates a session in the SessionManager (separate from PTY spawning). */
export async function createSession(
  id: number,
  mode: AiMode,
  projectPath: string
): Promise<SessionConfig> {
  return invoke<SessionConfig>("create_session", { id, mode, projectPath });
}

/** Assigns a branch and optional worktree path to a session. */
export async function assignSessionBranch(
  sessionId: number,
  branch: string,
  worktreePath: string | null
): Promise<void> {
  return invoke("assign_session_branch", { sessionId, branch, worktreePath });
}

/**
 * Subscribes to the per-session `pty-output-{sessionId}` Tauri event.
 * Returns a promise that resolves to an unlisten function. The caller must
 * invoke the unlisten function on cleanup to avoid leaked event listeners.
 */
export function onPtyOutput(
  sessionId: number,
  callback: (data: string) => void,
): Promise<UnlistenFn> {
  return listen<string>(`pty-output-${sessionId}`, (event) => {
    callback(event.payload);
  });
}

/** Backend info as returned by the Rust backend. */
export interface BackendInfo {
  backendType: BackendType;
  capabilities: BackendCapabilities;
}

/** Cached backend info to avoid repeated IPC calls. */
let cachedBackendInfo: BackendInfo | null = null;

/**
 * Returns information about the active terminal backend.
 * The result is cached after the first call.
 */
export async function getBackendInfo(): Promise<BackendInfo> {
  if (cachedBackendInfo) {
    return cachedBackendInfo;
  }
  cachedBackendInfo = await invoke<BackendInfo>("get_backend_info");
  return cachedBackendInfo;
}

/** Checks if the current backend supports enhanced terminal state. */
export async function hasEnhancedState(): Promise<boolean> {
  const info = await getBackendInfo();
  return info.capabilities.enhancedState;
}

File: src-tauri/src/commands/terminal.rs
use std::collections::HashMap;
use std::sync::Arc;

use serde::Serialize;
use tauri::{AppHandle, State};

use crate::core::session_manager::SessionManager;
use crate::core::status_server::StatusServer;
use crate::core::{BackendCapabilities, BackendType, ProcessManager, PtyError, SessionProcessTree};

/// Backend information returned to the frontend.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BackendInfo {
    /// The active backend type.
    pub backend_type: BackendType,
    /// Backend capabilities.
    pub capabilities: BackendCapabilitiesDto,
}

/// DTO for backend capabilities (frontend-friendly naming).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BackendCapabilitiesDto {
    pub enhanced_state: bool,
    pub text_reflow: bool,
    pub kitty_graphics: bool,
    pub shell_integration: bool,
    pub backend_name: String,
}

impl From<BackendCapabilities> for BackendCapabilitiesDto {
    fn from(caps: BackendCapabilities) -> Self {
        Self {
            enhanced_state: caps.enhanced_state,
            text_reflow: caps.text_reflow,
            kitty_graphics: caps.kitty_graphics,
            shell_integration: caps.shell_integration,
            backend_name: caps.backend_name.to_string(),
        }
    }
}

/// Returns information about the active terminal backend.
///
/// The frontend can use this to enable/disable features based on
/// backend capabilities (e.g., enhanced terminal state queries).
#[tauri::command]
pub fn get_backend_info() -> BackendInfo {
    let backend_type = BackendType::platform_default();

    let capabilities = match backend_type {
        BackendType::XtermPassthrough => BackendCapabilities {
            enhanced_state: false,
            text_reflow: false,
            kitty_graphics: false,
            shell_integration: false,
            backend_name: "xterm-passthrough",
        },
        BackendType::VteParser => BackendCapabilities {
            enhanced_state: true,
            text_reflow: false,
            kitty_graphics: false,
            shell_integration: false,
            backend_name: "vte-parser",
        },
    };

    BackendInfo {
        backend_type,
        capabilities: capabilities.into(),
    }
}

/// Exposes `ProcessManager::spawn_shell` to the frontend.
///
/// Validates that `cwd` (if provided) exists and is a directory before
/// forwarding to the process manager. Returns the new session ID.
/// The frontend should listen on `pty-output-{id}` for shell output events.
///
/// # Environment Variables
/// The `env` parameter allows passing environment variables to the shell process.
/// These are inherited by all child processes (including Claude CLI → MCP server).
/// Common usage: `{ "MAESTRO_PROJECT_HASH": "<hash>" }` for MCP status identification.
/// Note: `MAESTRO_SESSION_ID` is automatically set by the process manager.
#[tauri::command]
pub async fn spawn_shell(
    app_handle: AppHandle,
    state: State<'_, ProcessManager>,
    cwd: Option<String>,
    env: Option<HashMap<String, String>>,
) -> Result<u32, PtyError> {
    // Validate cwd if provided: must exist and be a directory
    let canonical_cwd = if let Some(ref dir) = cwd {
        let path = std::path::Path::new(dir);
        let canonical = path
            .canonicalize()
            .map_err(|e| PtyError::spawn_failed(format!("Invalid cwd '{dir}': {e}")))?;
        if !canonical.is_dir() {
            return Err(PtyError::spawn_failed(format!(
                "cwd '{dir}' is not a directory"
            )));
        }
        Some(canonical.to_string_lossy().into_owned())
    } else {
        None
    };
    let pm = state.inner().clone();
    pm.spawn_shell(app_handle, canonical_cwd, env)
}

/// Exposes `ProcessManager::write_stdin` to the frontend.
/// Sends raw text (including control sequences like `\r`) to the PTY.
#[tauri::command]
pub async fn write_stdin(
    state: State<'_, ProcessManager>,
    session_id: u32,
    data: String,
) -> Result<(), PtyError> {
    let pm = state.inner().clone();
    pm.write_stdin(session_id, &data)
}

/// Exposes `ProcessManager::resize_pty` to the frontend.
/// Rejects dimensions that are zero or exceed 500 to prevent misuse.
#[tauri::command]
pub async fn resize_pty(
    state: State<'_, ProcessManager>,
    session_id: u32,
    rows: u16,
    cols: u16,
) -> Result<(), PtyError> {
    if rows == 0 || cols == 0 || rows > 500 || cols > 500 {
        return Err(PtyError::resize_failed("Invalid dimensions"));
    }
    let pm = state.inner().clone();
    pm.resize_pty(session_id, rows, cols)
}

/// Exposes `ProcessManager::kill_session` to the frontend.
/// Gracefully terminates the PTY session (SIGTERM, then SIGKILL after 3s).
/// Also unregisters the session from the status server.
#[tauri::command]
pub async fn kill_session(
    state: State<'_, ProcessManager>,
    session_mgr: State<'_, SessionManager>,
    status_server: State<'_, Arc<StatusServer>>,
    session_id: u32,
) -> Result<(), PtyError> {
    // Kill the PTY session
    let pm = state.inner().clone();
    let result = pm.kill_session(session_id).await;

    // Unregister the session from the status server so it stops accepting updates
    status_server.unregister_session(session_id).await;

    // Log for debugging
    let _project_path = session_mgr
        .all_sessions()
        .into_iter()
        .find(|s| s.id == session_id)
        .map(|s| s.project_path);

    result
}

/// Returns the process tree for a specific session.
///
/// The tree includes the root shell process and all its descendants.
/// Returns None if the session doesn't exist or its root process has exited.
#[tauri::command]
pub async fn get_session_process_tree(
    state: State<'_, ProcessManager>,
    session_id: u32,
) -> Result<Option<SessionProcessTree>, String> {
    let pm = state.inner().clone();
    let root_pid = match pm.get_session_pid(session_id) {
        Some(pid) => pid,
        None => return Ok(None),
    };

    Ok(crate::core::process_tree::get_process_tree(session_id, root_pid))
}

/// Returns process trees for all active sessions.
///
/// More efficient than calling get_session_process_tree for each session
/// since it only refreshes the process list once.
#[tauri::command]
pub async fn get_all_process_trees(
    state: State<'_, ProcessManager>,
) -> Result<Vec<SessionProcessTree>, String> {
    let pm = state.inner().clone();
    let sessions = pm.get_all_session_pids();
    Ok(crate::core::process_tree::get_all_process_trees(&sessions))
}

/// Kills a specific process by PID.
///
/// Sends SIGTERM first, waits up to 2 seconds, then SIGKILL if still alive.
/// Will refuse to kill root session processes (use kill_session for that).
#[tauri::command]
pub async fn kill_process(
    state: State<'_, ProcessManager>,
    pid: u32,
) -> Result<(), String> {
    let pm = state.inner().clone();
    let session_root_pids: Vec<i32> = pm
        .get_all_session_pids()
        .into_iter()
        .map(|(_, root_pid)| root_pid)
        .collect();

    crate::core::process_tree::kill_process(pid, &session_root_pids)
        .await
        .map_err(|e| e.to_string())
}

/// Kills all active PTY sessions.
///
/// Used to clean up orphaned sessions when the frontend reloads.
/// Returns the number of sessions that were killed.
#[tauri::command]
pub async fn kill_all_sessions(state: State<'_, ProcessManager>) -> Result<u32, PtyError> {
    let pm = state.inner().clone();
    pm.kill_all_sessions().await
}

/// Checks if a command is available in the user's PATH.
/// Uses platform-appropriate method:
/// - Unix: runs `command -v <cmd>` via interactive login shell to get user's real PATH
/// - Windows: runs `where.exe <cmd>`
#[tauri::command]
pub async fn check_cli_available(command: String) -> Result<bool, String> {
    #[cfg(unix)]
    {
        let shell = std::env::var("SHELL").unwrap_or_else(|_| "/bin/zsh".to_string());

        // First, get the user's real PATH from their shell profile
        // This handles nvm, homebrew, etc. that modify PATH in .zshrc/.bashrc
        let path_output = tokio::process::Command::new(&shell)
            .args(["-l", "-i", "-c", "echo $PATH"])
            .output()
            .await
            .map_err(|e| format!("Failed to get PATH: {}", e))?;

        let user_path = String::from_utf8_lossy(&path_output.stdout)
            .trim()
            .to_string();

        // Now check for the command using the user's PATH
        let output = tokio::process::Command::new(&shell)
            .args(["-l", "-c", &format!("command -v {}", command)])
            .env("PATH", &user_path)
            .output()
            .await
            .map_err(|e| format!("Failed to check CLI: {}", e))?;

        Ok(output.status.success())
    }

    #[cfg(windows)]
    {
        let output = tokio::process::Command::new("where.exe")
            .arg(&command)
            .output()
            .await
            .map_err(|e| format!("Failed to check CLI: {}", e))?;
        Ok(output.status.success())
    }
}
